lisp/ess-r-xref.el

```diff
@@ -0,0 +1,163 @@
+
+Most functions will not have source file references, either
+because they were sent to the interpreter directly, or because
+they are loaded via byte-compiled packages."
+  (ess-boolean-command
+   (format
+    "!is.null(utils::getSrcref(%s)) && file.exists(utils::getSrcFilename(%s))\n"
+    symbol symbol)))
+
+(defun ess-r-xref--find-srcfile (symbol)
+  "Creates an xref for the source file reference of R symbol SYMBOL."
+  (let ((file (ess-string-command
+               (format "cat(utils::getSrcFilename(%s)[1], \"\\n\")\n" symbol)))
+        (line (ess-string-command
+               (format "cat(utils::getSrcLocation(%s, which = \"line\")[1], \"\\n\")\n" symbol))))
+    (list (xref-make (format "%s" symbol)
+                     (xref-make-file-location
+                      (expand-file-name (string-trim file))
+                      (string-to-number line)
+                      0)))))
+
+
+;;; Local/Byte-compiled Locations
+
+(defun ess-r-xref--is-function (symbol)
+  "Check whether SYMBOL is an R symbol."
+  (ess-boolean-command (format "is.function(%s)\n" symbol)))
+
+(defun ess-r-xref--find-body (fn)
+  "Creates an xref to a buffer containing the body of the R function FN."
+  (let ((cmd (format "cat(\"%s <- \"); print.function(%s)\n"
+                     fn fn))
+        (buff (get-buffer-create
+               (format "*definition[R]:%s*" fn)))
+        (proc ess-local-process-name))
+    (with-current-buffer (ess-command cmd buff)
+      (ess-r-source-mode)
+      (setq-local ess-local-process-name proc))
+    (list (xref-make (format "%s" fn) (xref-make-buffer-location buff 0)))))
+
+
+;;; R Function Source Viewer Mode
+
+(defun ess-r-xref--parse-byte-compiled-p (buff)
+  "Determine whether the R function printed in BUFF is byte-compiled.
+
+This also erases the metadata once the compilation status has
+been determined."
+  (with-current-buffer buff
+    (goto-char (point-max))
+    (when (re-search-backward "^<bytecode" nil 'noerror)
+      ;; Delete the metadata, if it's there.
+      (beginning-of-line)
+      (kill-whole-line)
+      t)))
+
+(defun ess-r-xref--parse-environment (buff)
+  "Determine the environment of the R function printed in BUFF.
+
+This also erases the metadata once the environment has been
+determined."
+  (with-current-buffer buff
+    (goto-char (point-max))
+    (let* ((found (re-search-backward "^<environment: \\(.*\\)>$" nil 'noerror))
+           (env (or (when found (match-string-no-properties 1)) ".GlobalEnv")))
+      ;; Delete the metadata, if it's there.
+      (when found (beginning-of-line) (kill-whole-line))
+      env)))
+
+;; Defined in ess-r-mode.
+(defvar ess-r-syntax-table)
+
+(defvar-local ess-r-source-byte-compiled-p nil
+  "Indicates whether the R function on display is byte-compiled.")
+
+(defvar-local ess-r-source-environment nil
+  "The environment the R function on display belongs to.")
+
+(define-derived-mode ess-r-source-mode special-mode "R Source"
```

Comment:
Why is this mode needed? (I cannot test the code at the moment and I don't have a clear picture of what's going on.)

Reply:
Although it shares syntax with R buffers, it's not an editable buffer (since it represents the internal state of the R interpreter), and so it doesn't need any of the usual ESS apparatus (evaluation, completion, etc). I thought making a simple special mode was a good way to express this.

Reply:
Only if that package has been `source()`'d into a running interpreter. Packages that are byte-compiled or evaluated linewise into the interpreter do not have source files known to the R process, so the best we can do is dump out what R knows (i.e. print the function). If the function is not known to a running R process at all, the jump will simply fail, or fall back onto etags (or whatever else you have set in your `xref-backend-functions` list).

Reply:
@atheriel : Note that byte-compiled code will be come prevalent in R in the future.  *But* that byte-compiled code will still have source references (and has now).  So I hope you assume source references but you do*not* assume non-bytecompiled code.

Reply:
@mmaechler Byte-compiled functions do not appear to have the `srcref` attribute, at least in my own testing -- is there somewhere else the source location is stored?

Reply:
Only locally defined functions will have srcrefs. But my guess is that you don't make any distinction between bytecoded and non-bytecoded functions and just look at the srcref attribute? It would be hard to make the distinction anyway, they have the same type.

Reply:
The distinction is made by parsing the output of `print.function()`. If there's a bytecode string at the bottom, it's considered byte-compiled, and otherwise it is not. I don't actually do anything with this information, though, other than put it in the header.